internal/flowcontrol/writebuffer.go

```diff
@@ -198,54 +172,70 @@ func (w *ResourceSliceWriteBuffer) updateSlice(ctx context.Context, sliceNSN typ
 		return false
 	}
 
-	logger.V(0).Info(fmt.Sprintf("updated the status of %d resources in slice", len(updates)))
+	logger.V(0).Info(fmt.Sprintf("updated the status of %d resources in slice", len(updates)), "latency", time.Since(insertionTime).Abs().Milliseconds())
 	sliceStatusUpdates.Inc()
 	return true
 }
 
-type jsonPatch struct {
-	Op    string `json:"op"`
-	Path  string `json:"path"`
-	Value any    `json:"value"`
-}
-
-type rateLimiter struct {
-	failuresLock sync.Mutex
-	failures     map[interface{}]int
-	limiter      *rate.Limiter
-}
+func (*ResourceSliceWriteBuffer) buildPatch(slice *apiv1.ResourceSlice, updates []*resourceSliceStatusUpdate) []*jsonPatch {
+	var patches []*jsonPatch
+	unsafeSlice := slice.Status.Resources
 
-func newRateLimiter(batchInterval time.Duration, burst int) workqueue.RateLimiter {
-	return &rateLimiter{
-		failures: map[interface{}]int{},
-		limiter:  rate.NewLimiter(rate.Every(batchInterval), burst),
+	if len(unsafeSlice) == 0 {
+		patches = append(patches,
+			&jsonPatch{
+				Op:    "add",
+				Path:  "/status",
+				Value: map[string]any{},
+			},
+			&jsonPatch{
+				Op:    "test",
+				Path:  "/status/resources",
+				Value: nil,
+			},
+			&jsonPatch{
+				Op:    "add",
+				Path:  "/status/resources",
+				Value: make([]apiv1.ResourceState, len(slice.Spec.Resources)),
+			})
 	}
-}
-
-func (r *rateLimiter) When(item any) time.Duration {
-	r.failuresLock.Lock()
-	defer r.failuresLock.Unlock()
 
-	failures := r.failures[item]
-	r.failures[item]++
+	for _, update := range updates {
+		if update.SlicedResource.Index > len(slice.Spec.Resources)-1 || update.SlicedResource.Index < 0 {
```

Comment:
Any reason to have it if it's impossible? And if it's just highly improbable, should we log?

Reply:
If this somehow did evaluate to true in production, I worry that it would generate a ton of logs. I'm personally okay with swallowing this condition but open to other options

Reply:
Swallowing is fine for now, though I wonder if the fact that it was true in prod reflects on a deeper issue we're not seeing